### PR TITLE
[DS-1657] view post response engine code 

### PR DESCRIPTION
**Pull request by aditya-sunil-flipkart**

### CODE REVIEW

File: moriarty-core/src/main/java/com/flipkart/sherlock/helper/ViewHelper.java [hunk 1]
```diff
@@ -59,7 +61,7 @@ public String getFqField(StoreView storeView) {
     }
 
     public String getFacetField(StoreView storeView) {
-        return solrFieldHelper.getSolrField(storeView.getField(), storeView.getField()) + getFqTypeSuffix(storeView.getFieldType());
+        return solrFieldHelper.getSolrField(storeView.getField(), storeView.getField()) + getFacetTypeSuffix(storeView.getFieldType());
```
Comment (aditya-sunil-flipkart) [inline moriarty-core/src/main/java/com/flipkart/sherlock/helper/ViewHelper.java]: bug in the previously written code..

File: moriarty-core/src/main/java/com/flipkart/sherlock/engine/view/ViewPostResponseEngine.java [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+
+        String queryParamString = createQueryParamString(moriartyRequest, replacementMap);
+        String uri = moriartyRequest.getUri() + queryParamString;
+
+        resource.setUri(uri);
+
+        return resource;
+    }
+
+    private String createQueryParamString(MoriartyRequest moriartyRequest, Map<String, String> replacementMap) {
```
Comment (setupoddar) [inline moriarty-core/src/main/java/com/flipkart/sherlock/engine/view/ViewPostResponseEngine.java]: I am not sure what you achieved with this method, I dont see any diff between 
replacedQueryParams and moriartyRequest.getQueryParameters()
For me it is the same.

File: moriarty-core/src/main/java/com/flipkart/sherlock/engine/view/ViewPostResponseEngine.java [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+        }
+        return URIUtils.getURIFromMap(replacedQueryParams);
+    }
+
+    @Override
+    public String getEngineName() {
+        return VIEW_POST_RESPONSE_ENGINE;
+    }
+}
+
```
Comment (setupoddar) [inline moriarty-core/src/main/java/com/flipkart/sherlock/engine/view/ViewPostResponseEngine.java]: remove the extra line

File: moriarty-core/src/main/java/com/flipkart/sherlock/helper/ViewHelper.java (hunk 1 above)
Comment (setupoddar) [reply moriarty-core/src/main/java/com/flipkart/sherlock/helper/ViewHelper.java]: thnx

File: moriarty-core/src/main/java/com/flipkart/sherlock/engine/view/ViewPostResponseEngine.java [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+                String facetField = viewHelper.getFacetField(storeView);
+                if (facetFields.containsKey(facetField)) {
+                    Map<String, Long> facetValues = (Map<String, Long>) facetFields.get(facetField);
+                    if (facetValues.containsKey(viewParam)) {
+                        return title.replaceAll(Pattern.quote("{TITLE}"), viewParam) ;
+                    }
+                }
+            }
+        }
+        return title.replaceAll(Pattern.quote("{TITLE}"), "");
```
Comment (setupoddar) [inline moriarty-core/src/main/java/com/flipkart/sherlock/engine/view/ViewPostResponseEngine.java]: Use StringUtils.Empty

Comment (setupoddar) [inline moriarty-core/src/main/java/com/flipkart/sherlock/engine/view/ViewPostResponseEngine.java]: All add constant for {TITLE}

File: moriarty-core/src/main/java/com/flipkart/sherlock/engine/view/ViewPostResponseEngine.java [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+
+    private ViewHelper viewHelper;
+
+    public ViewPostResponseEngine(ViewHelper viewHelper) {
+        this.viewHelper = viewHelper;
+    }
+
+    @Override
+    public boolean applyPredicate(OrchestratorContext orchestratorContext) {
+        return viewHelper.isApplicable(orchestratorContext) && MoriartyEngineUtil.getMoriartyRequest(orchestratorContext).hasMetadata("view");
```
[the remark this answers is not in the document]
Comment (aditya-sunil-flipkart) [reply moriarty-core/src/main/java/com/flipkart/sherlock/engine/view/ViewPostResponseEngine.java]: done!

File: moriarty-core/src/main/java/com/flipkart/sherlock/engine/view/ViewPostResponseEngine.java [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+                }
+            }
+        }
+        return title.replaceAll(Pattern.quote("{TITLE}"), "");
+    }
+
+    private Map<String, Object> metadataAdaptor(Set<MetaData> metaDataSet) {
+        Map<String, Object> metadataMap = new HashMap<>();
+        for (MetaData metaData: metaDataSet) {
+            metadataMap.put(metaData.getName(), metaData.getValue());
```
[the remark this answers is not in the document]
Comment (aditya-sunil-flipkart) [reply moriarty-core/src/main/java/com/flipkart/sherlock/engine/view/ViewPostResponseEngine.java]: done!

File: moriarty-core/src/main/java/com/flipkart/sherlock/engine/view/ViewPostResponseEngine.java [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+        }
+        return metadataMap;
+    }
+
+    private Resource createResource(MoriartyRequest moriartyRequest, Map<String, String> replacementMap) {
```
[the remark this answers is not in the document]
Comment (aditya-sunil-flipkart) [reply moriarty-core/src/main/java/com/flipkart/sherlock/engine/view/ViewPostResponseEngine.java]: wanted to make a generalized method where multiple params can be passed, hence passed a map.
Agreed on the second point. Also will move the helper methods to viewHelper class.

File: moriarty-core/src/main/java/com/flipkart/sherlock/engine/view/ViewPostResponseEngine.java (hunk 2 above)
Comment (aditya-sunil-flipkart) [reply moriarty-core/src/main/java/com/flipkart/sherlock/engine/view/ViewPostResponseEngine.java]: This method is taking keys from moriartyRequest.getQueryParameters() , if it is present in replacementMap , value from replacementMap is put in the final string and value from moriartyRequest is ignored. 
Yes but in this case, this method is not doing anything as view.id & view.param are same as original query.

File: moriarty-core/src/main/java/com/flipkart/sherlock/engine/view/ViewPostResponseEngine.java (hunk 3 above)
Comment (aditya-sunil-flipkart) [reply moriarty-core/src/main/java/com/flipkart/sherlock/engine/view/ViewPostResponseEngine.java]: done!

File: moriarty-core/src/main/java/com/flipkart/sherlock/engine/view/ViewPostResponseEngine.java (hunk 4 above)
Comment (aditya-sunil-flipkart) [reply moriarty-core/src/main/java/com/flipkart/sherlock/engine/view/ViewPostResponseEngine.java]: done!

Comment (aditya-sunil-flipkart) [reply moriarty-core/src/main/java/com/flipkart/sherlock/engine/view/ViewPostResponseEngine.java]: done!

File: moriarty-core/src/main/java/com/flipkart/sherlock/helper/ViewHelper.java [hunk 8]
```diff
@@ -133,4 +143,80 @@ public Query getQuery(StoreView view) {
...
+        if (searchEngineResponse != null) {
+            ResponseObject responseObject = searchEngineResponse.getResponse();
+            Map<String, Object> facet_counts = (Map<String, Object>) responseObject.getResponse().get("facet_counts");
+            if (facet_counts != null && facet_counts.containsKey("facet_fields")) {
+                Map<String, Object> facetFields = (Map<String, Object>) facet_counts.get("facet_fields");
+                String facetField = getFacetField(storeView);
+                if (facetFields.containsKey(facetField)) {
+                    Map<String, Long> facetValues = (Map<String, Long>) facetFields.get(facetField);
+                    if (facetValues.containsKey(viewParam)) {
+                        return title.replaceAll(Pattern.quote("{TITLE}"), viewParam) ;
```
Comment (setupoddar) [inline moriarty-core/src/main/java/com/flipkart/sherlock/helper/ViewHelper.java]: Why not use ViewResponseEngine.TITLE

File: moriarty-core/src/main/java/com/flipkart/sherlock/helper/ViewHelper.java [hunk 9]
```diff
@@ -133,4 +143,80 @@ public Query getQuery(StoreView view) {
...
+            while (iterator.hasNext()) {
+                Map.Entry<String, String> next = iterator.next();
+                resourceParam.append(next.getKey()).append("=").append(next.getValue());
+                if (iterator.hasNext()) {
+                    resourceParam.append("&");
+                }
+            }
+        }
+        else {
+            resource.setParams("");
```
Comment (setupoddar) [inline moriarty-core/src/main/java/com/flipkart/sherlock/helper/ViewHelper.java]: EMPTY String Constant

File: moriarty-core/src/main/java/com/flipkart/sherlock/helper/ViewHelper.java [hunk 10]
```diff
@@ -133,4 +143,80 @@ public Query getQuery(StoreView view) {
...
+
+        String queryParamString = createQueryParamString(moriartyRequest, replacementMap, true);
+        String uri = moriartyRequest.getUri() + queryParamString;
+
+        resource.setUri(uri);
+
+        return resource;
+    }
+
+    public String createQueryParamString(MoriartyRequest moriartyRequest, Map<String, String> replacementMap, boolean isSelected) {
```
Comment (setupoddar) [inline moriarty-core/src/main/java/com/flipkart/sherlock/helper/ViewHelper.java]: You pointed out your self that we are not doing anything here in this method, then why this method?

File: moriarty-core/src/main/java/com/flipkart/sherlock/helper/ViewHelper.java [hunk 11]
```diff
@@ -127,4 +139,84 @@ public Query getQuery(StoreView view) {
...
+        return title.replaceAll(ViewResponseEngine.TITLE, StringUtils.EMPTY);
+    }
+
+    public Resource createResource(MoriartyRequest moriartyRequest, Map<String, String> replacementMap) {
+        Resource resource = new Resource();
+        resource.setSelected(true);
+
+        StringBuilder resourceParam = new StringBuilder();
+
+        if (!replacementMap.isEmpty()) {
```
Comment (setupoddar) [inline moriarty-core/src/main/java/com/flipkart/sherlock/helper/ViewHelper.java]: Use MapUtils

File: moriarty-core/src/main/java/com/flipkart/sherlock/helper/ViewHelper.java [hunk 12]
```diff
@@ -127,4 +139,84 @@ public Query getQuery(StoreView view) {
...
+                if (iterator.hasNext()) {
+                    resourceParam.append("&");
+                }
+            }
+        }
+        else {
+            resource.setParams("");
+        }
+
+        String queryParamString = createQueryParamString(moriartyRequest, replacementMap, true);
```
Comment (setupoddar) [inline moriarty-core/src/main/java/com/flipkart/sherlock/helper/ViewHelper.java]: These 3 lines can be written in 1 line.

**Pull request merged**